Lint with clippy in Travis builds

oliverlee opened the pull request.

mickvangelderen commented on .travis.yml:
```diff
@@ -16,11 +16,13 @@ before_script:
   - if [ "$TRAVIS_RUST_VERSION" == "nightly" ]; then
         rustup component add --toolchain nightly rustfmt-preview;
```
I'm not sure but isn't `rustfmt` currently preferred over the older `rustfmt-preview`.

oliverlee replied:
I think I just copied some old travis script so I'll change that.

oliverlee replied:
Apparently if you look at https://static.rust-lang.org/dist/channel-rust-nightly.toml, it renames rustfmt to rustfmt-preview.

```
oliver@canopus:~/repos/dominion$  curl -s https://static.rust-lang.org/dist/channel-rust-nightly.toml | grep -A1 rename | grep rustfmt
[renames.rustfmt]
to = "rustfmt-preview"
```

mickvangelderen commented on .travis.yml:
```diff
@@ -16,11 +16,13 @@ before_script:
   - if [ "$TRAVIS_RUST_VERSION" == "nightly" ]; then
         rustup component add --toolchain nightly rustfmt-preview;
         which rustfmt || cargo install --force rustfmt-nightly;
+        rustup component add clippy;
```
Does it matter that not all nightly versions support clippy? Can the build fail if thats the case (quite common: https://rust-lang.github.io/rustup-components-history/index.html)

oliverlee replied:
Can we make a stable job that only runs clippy? Or will that fail due to use of nightly features?

oliverlee replied:
https://github.com/rust-lang/rust-clippy/#travis-ci

oliverlee committed:
Lint with clippy in Travis

Create a new job that lints with rustfmt and clippy.

This commit also removes the default-run feature as it is now stable and
is no longer necessary to be listed in the manifest.

oliverlee committed:
Fix simple clippy lints

This commit fixes the following lints:
- unreachable_patterns
- unused_imports
- unused_variables
- clippy::collapsible_if
- clippy::into_iter_on_ref
- clippy::len_zero
- clippy::let_unit_value
- clippy::match_ref_pats
- clippy::option_map_unit_fn
- clippy::print_with_newline
- clippy::redundant_static_lifetimes
- clippy::trivially_copy_pass_by_ref
- clippy::unnecessary_operation

oliverlee committed:
Fix clippy::new_without_default in Turn

oliverlee committed:
Suppress dead_code warnings

oliverlee committed:
Fix clippy::ptr_arg in handling effects

oliverlee committed:
Deny clippy::pedantic lints

Allow lint clippy::module_name_repetitions for card::CardKind and
kingdom::KingdomSet as they are commonly used without the {card,kingdom}
modules. Allow lint clippy::match_same_arms for generated CardKind
methods.

Allow lint clippy::non_ascii_literal in scrape.rs for matching Potion
symbols.
